src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -343,26 +394,52 @@ private bool IsRecovering(ServerErrorCode code, string message)
             return false;
         }
 
+        private bool IsRecoveringErrorException(Exception exception)
+        {
+            return
+                exception is MongoCommandException commandException &&
+                IsRecovering((ServerErrorCode)commandException.Code, commandException.ErrorMessage);
+        }
+
+        private void SetDescription(ServerDescription newDescription)
+        {
+            var oldDescription = Interlocked.CompareExchange(ref _currentDescription, newDescription, _currentDescription);
```

Comment:
See this [comment thread ](https://github.com/vincentkam/mongo-csharp-driver/pull/115#discussion_r430476588)for the context behind why I chose to use a slightly different pattern from the one in `ServerMonitor`